ibm/resource_ibm_network_gateway.go

```diff
@@ -0,0 +1,996 @@
+		Pending: []string{"retry", "active"},
+		Target:  []string{"idle"},
+		Refresh: func() (interface{}, string, error) {
+			bm, err := service.Id(id).Mask("id,activeTransactionCount").GetObject()
+			if err != nil {
+				return false, "retry", nil
+			}
+
+			if bm.ActiveTransactionCount != nil && *bm.ActiveTransactionCount == 0 {
+				return bm, "idle", nil
+			}
+			return bm, "active", nil
+
+		},
+		Timeout:        24 * time.Hour,
+		Delay:          10 * time.Second,
+		MinTimeout:     1 * time.Minute,
+		NotFoundChecks: 24 * 60,
+	}
+
+	return stateConf.WaitForState()
+}
+
+// Network gateways or Bare metal creation does not return a  object with an Id.
+// Have to wait on provision date to become available on server that matches
+// hostname and domain.
+// http://sldn.softlayer.com/blog/bpotter/ordering-bare-metal-servers-using-softlayer-api
+func waitForNetworkGatewayProvision(d *datatypes.Hardware, meta interface{}) (interface{}, error) {
+	hostname := *d.Hostname
+	domain := *d.Domain
+	log.Printf("Waiting for Gateway (%s.%s) to be provisioned", hostname, domain)
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"retry", "pending"},
+		Target:  []string{"provisioned"},
+		Refresh: func() (interface{}, string, error) {
+			service := services.GetAccountService(meta.(ClientSession).SoftLayerSession())
+			bms, err := service.Filter(
+				filter.Build(
+					filter.Path("hardware.hostname").Eq(hostname),
+					filter.Path("hardware.domain").Eq(domain),
+				),
+			).Mask("id,provisionDate").GetHardware()
+			if err != nil {
+				return false, "retry", nil
+			}
+
+			if len(bms) == 0 || bms[0].ProvisionDate == nil {
+				return datatypes.Hardware{}, "pending", nil
+			} else {
+				return bms[0], "provisioned", nil
+			}
+		},
+		Timeout:        24 * time.Hour,
+		Delay:          10 * time.Second,
+		MinTimeout:     1 * time.Minute,
+		NotFoundChecks: 24 * 60,
+	}
+
+	return stateConf.WaitForState()
+}
+
+func findMemoryItemPriceIdGateway(items []datatypes.Product_Item, d *schema.ResourceData) (datatypes.Product_Item_Price, error) {
+	memory := d.Get("memory").(int)
+	memoryStr := "RAM_" + strconv.Itoa(memory) + "_GB"
+	availableMemories := ""
+
+	for _, item := range items {
+		for _, itemCategory := range item.Categories {
+			if *itemCategory.CategoryCode == "ram" {
+				availableMemories = availableMemories + *item.KeyName + "(" + *item.Description + ")" + ", "
+				if strings.HasPrefix(*item.KeyName, memoryStr) {
+					for _, price := range item.Prices {
+						if price.LocationGroupId == nil {
+							return datatypes.Product_Item_Price{Id: price.Id}, nil
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return datatypes.Product_Item_Price{},
+		fmt.Errorf("Could not find the price item for %d GB memory. Available items are %s", memory, availableMemories)
+}
+func findNetworkItemPriceIdGateway(items []datatypes.Product_Item, d *schema.ResourceData) (datatypes.Product_Item_Price, error) {
```

Comment:
Can't we re-use `findNetworkItemPriceId` from `ibm_bare_metal` implementation?

Reply:
All instances of findNetworkItemPriceIdGateway have been changed to findNetworkItemPriceId